snuba/datasets/querylog.py

```diff
@@ -0,0 +1,92 @@
+from datetime import timedelta
+from typing import Mapping
+
+from snuba.clickhouse.columns import (
+    Array,
+    ColumnSet,
+    DateTime,
+    LowCardinality,
+    Nested,
+    Nullable,
+    String,
+    UInt,
+    UUID,
+)
+
+from snuba.datasets.dataset import Dataset
+from snuba.datasets.dataset_schemas import DatasetSchemas
+from snuba.datasets.querylog_processor import QueryLogMessageProcessor
+from snuba.datasets.schemas.tables import MergeTreeSchema
+from snuba.datasets.table_storage import TableWriter, KafkaStreamLoader
+from snuba.query.extensions import QueryExtension
+from snuba.query.timeseries import TimeSeriesExtension
+
+
+class QueryLogDataset(Dataset):
+    """
+    Record a sample of the queries ran on snuba with stats useful to do some
+    performance analysis.
+
+    This is meant to consume the queries topic populated by Snuba itself.
+    """
+
+    def __init__(self) -> None:
+        columns = ColumnSet(
+            [
+                ("clickhouse_query", String()),
+                ("snuba_query", String()),
+                ("timestamp", DateTime()),
+                ("duration_ms", UInt(32)),
+                ("referrer", String()),
+                # This is to assess the statistical accuracy if we sample the writes.
+                # hits = 1 / sampling_rate. Thus it means the amount of queries this row
+                # represents, and, as a consequence the accuracy (the higher hits, the lower
+                # the accuracy).
+                ("hits", UInt(32)),
+                # The trace id this was run into. This is meant to be used to match a query
+                # with ther APM data.
+                ("trace_id", UUID()),
+                # Details about the query
+                ("dataset", LowCardinality(String())),
+                ("from_clause", LowCardinality(String())),
+                ("columns", Array(LowCardinality(String()))),
+                ("limit", Nullable(UInt(32))),
+                ("offset", Nullable(UInt(32))),
+                # Further query specific fields (This should eventually be
+                # removed in favor of columns once we are comfortable on the schema)
+                ("tags", Nested([("key", String()), ("value", String())])),
+            ]
+        )
+
+        schema = MergeTreeSchema(
```

Comment:
@JTCunning, Ideally we would not sample the data we store to make it easy tracking down individual queries, but I guess the amount of data we would store is would be fairly high, so there is a sampling rate parameter in settings to reduce the load on clickhouse.

Rough math:
2K per event according to Clickhouse
500 writes per second 
this is 84 Gb per day. Which I think is too much.

What we can do is to sample this data and keep it for a short retention. And have a stripped down version of the table with less data and a longer retention.
Also I think this should be a distributed table for the high write rate.

What do you think?